src/vs/platform/quickinput/browser/quickPickPin.ts

```diff
@@ -0,0 +1,86 @@
+		if (e.button.iconClass === ThemeIcon.asClassName(Codicon.pin) || e.button.iconClass === ThemeIcon.asClassName(Codicon.pinned)) {
+			_formatPinnedItems(storageKey, quickPick, storageService, e, callback);
+			callback();
+		}
+	});
+	_formatPinnedItems(storageKey, quickPick, storageService, undefined, callback);
+}
+
+async function _formatPinnedItems(storageKey: string, quickPick: IQuickPick<IQuickPickItem>, storageService: IStorageService, event?: IQuickPickItemButtonEvent<IQuickPickItem>, callback?: () => Promise<void>): Promise<void> {
+	const formattedItems: QuickPickItem[] = [];
+	const labels = getPinnedLabels(storageKey, storageService);
+	const updatedLabels = !!event?.item.label ? updatePinned(storageKey, event.item.label, storageService, new Set(labels).has(event.item.label)) : labels.filter(l => l !== 'Pinned');
+	if (updatedLabels.length) {
+		formattedItems.push({ type: 'separator', label: localize("terminal.commands.pinned", 'Pinned') });
+	}
+	for (const label of updatedLabels) {
+		const item = quickPick.items.find(i => i.label === label && i.type !== 'separator');
+		if (item) {
+			const pinnedItem = Object.assign({}, item);
+			updateButtons(pinnedItem, false);
+			formattedItems.push(pinnedItem);
+		}
+	}
+
+	for (const item of quickPick.items.filter(i => !!i.label)) {
+		updateButtons(item, true);
+		formattedItems.push(item);
+	}
+	quickPick.items = formattedItems;
+	quickPick.onDidChangeValue(async value => {
+		// don't show pinned items in the search results
+		quickPick.items = value ? quickPick.items.filter(i => i.type !== 'separator' && !i.buttons?.find(b => b.iconClass === ThemeIcon.asClassName(Codicon.pinned))) : quickPick.items;
+		if (!value && callback) {
+			await callback();
+		}
+	});
+	quickPick.show();
+}
+
+function updateButtons(item: QuickPickItem, removePin: boolean): void {
+	if (item.type === 'separator') {
+		return;
+	}
+	item.buttons = item.buttons ? item.buttons?.filter(b => b.iconClass !== ThemeIcon.asClassName(Codicon.pin) && b.iconClass !== ThemeIcon.asClassName(Codicon.pinned)) : [];
+	item.buttons.unshift({
+		iconClass: removePin ? ThemeIcon.asClassName(Codicon.pin) : ThemeIcon.asClassName(Codicon.pinned),
+		tooltip: removePin ? localize('pinCommand', "Pin command") : localize('pinnedCommand', "Pinned command"),
+		alwaysVisible: false
+	});
+}
+
+function updatePinned(storageKey: string, label: string, storageService: IStorageService, removePin: boolean): string[] {
+	const pinnedLabels = getPinnedLabels(storageKey, storageService).filter(l => l !== label);
+	if (!removePin) {
+		pinnedLabels.push(label);
+	}
+	storageService.store(storageKey, JSON.stringify(pinnedLabels), StorageScope.WORKSPACE, StorageTarget.USER);
+	return pinnedLabels;
+}
+
+function getPinnedLabels(storageKey: string, storageService: IStorageService): string[] {
+	const items = storageService.get(storageKey, StorageScope.WORKSPACE);
+	return items ? JSON.parse(items) : [];
+}
```

Comment:
This should return an `IQuickPickItem[]` because I can totally see someone wanting to pin QPIs with details, descriptions... heck even buttons.

Reply:
Also, feel free to leverage the `id` property on `IQuickPickItem` for determining if an item already is pinned or not.